File: opi-core/src/main/java/org/lei/opi/core/O900.java
package org.lei.opi.core;

import java.util.HashMap;

import org.lei.opi.core.structures.Parameter;
import org.lei.opi.core.structures.ReturnMsg;

/**
 * Octopus O900 client
 *
 * @since 0.0.1
 */
public class O900 extends OpiMachine {

  private enum Eye {LEFT, RIGHT};
  private enum BgLum {BG_OFF, BG_1, BG_10, BG_100};
  private enum BgCol {WHITE, YELLOW};
  private enum FixType {CENTER, CROSS, RING};
  private enum StCol {WHITE, RED, BLUE};
  private enum StSize {GI, GII, GIII, GIV, GV, GVI};

  /**
   * Octopus O900 constructor
   *
   * @since 0.0.1
   */
  public O900() { super(); }

  /**
   * opiQuery: Query device
   * 
   * @return settings and state machine state
   *
   * @since 0.0.1
   */
  @ReturnMsg(name = "error", desc = "Empty string for all good, else error message.")
  @ReturnMsg(name = "msg", desc = "JSON Object with all of the other fields described in @ReturnMsg except 'error'.")
  @ReturnMsg(name = "msg.jovp", desc = "Any messages that the JOVP sent back.")
   public MessageProcessor.Packet query() {
    return new MessageProcessor.Packet("");
  }

  /**
   * opiInitialise: initialize OPI
   * 
   * @param args A map of name:value pairs for Params
   * 
   * @return A JSON object with return messages
   * 
   * @since 0.0.1
   */
  @Parameter(name = "ip", desc = "IP Address of the perimeter.", defaultValue = "192.126.0.1")
  @Parameter(name = "port", desc = "TCP port of the perimeter.", className = Double.class, min = 0, max = 65535, defaultValue = "50000")
  @Parameter(name = "ip_Monitor", desc = "IP Address of the O900 server.", defaultValue = "localhost")
  @Parameter(name = "port_Monitor", desc = "TCP port of the O900 server.", className = Double.class, min = 0, max = 65535, defaultValue = "50001")
  @Parameter(name = "eye", desc = "Eye to set.", className = Eye.class, defaultValue = "left")
  @Parameter(name = "eyeSuite", desc = "Path to EyeSuite.", className = String.class, defaultValue = "C:/XXX/eyeSuite/")
  @Parameter(name = "gazeFeed", desc = "Path where to save gaze feed. Directory must exists", className = String.class, defaultValue = "C:/XXX/gazeFeed/")
  @Parameter(name = "bigWheel", desc = "Whether O900 has a big wheel for displaying Goldmann Size VI stimuli.", className = Double.class, min = 0, max = 1, defaultValue = "0")
  @Parameter(name = "pres", desc = "Volume for auditory feedback when a stimulus is presented: 0 means no buzzer.", className = Double.class, min = 0, max = 3, defaultValue = "0")
  @Parameter(name = "resp", desc = "Volume for auditory feedback when observer presses the clicker: 0 means no buzzer.", className = Double.class, min = 0, max = 3, defaultValue = "0")
  @Parameter(name = "max10000", desc = "Whether O900 can handle a maximum luminance of 10000 apostilbs instead of 4000. Check the settings in EyeSuite", className = Double.class, min = 0, max = 1, defaultValue = "0")
  @ReturnMsg(name = "error", desc = "Empty string for all good, else error messages from Imo.")
  @ReturnMsg(name = "msg", desc = "JSON Object with all of the other fields described in @ReturnMsg except 'error'.")
  @ReturnMsg(name = "msg.jovp", desc = "Any messages that the JOVP sent back.")
  public MessageProcessor.Packet initialize(HashMap<String, Object> args) {
    setIsInitialised(true);
    return new MessageProcessor.Packet("");
  }

  /**
   * opiSetup: Change device background and overall settings
   * 
   * @param args pairs of argument name and value
   * 
   * @return A JSON object with return messages
   *
   * @since 0.0.1
   */
  @Parameter(name = "bgLum", desc = "Background luminance for eye.", className = BgLum.class, defaultValue = "white")
  @Parameter(name = "bgCol", desc = "Background color for eye.", className = BgCol.class, defaultValue = "white")
  @Parameter(name = "fixType", desc = "Fixation target type for eye.", className = FixType.class, defaultValue = "center")
  @Parameter(name = "fixLum", desc = "Fixation luminance color for eye (from 0% to 100%).", className = Double.class, defaultValue = "50", min = 0, max = 100)
  @Parameter(name = "f310", desc = "Whether to use Logitech's F310 controlles", className = Double.class, defaultValue = "0", min = 0, max = 1)
  @ReturnMsg(name = "error", desc = "Empty string for all good, else error messages from ImoVifa.")
  @ReturnMsg(name = "msg", desc = "JSON Object with all of the other fields described in @ReturnMsg except 'error'.")
  @ReturnMsg(name = "msg.jovp", desc = "Any messages that the JOVP sent back.")
  public MessageProcessor.Packet setup(HashMap<String, Object> args) {
    return new MessageProcessor.Packet("");
  }

  /**
   * opiPresent: Present OPI stimulus in perimeter
   * 
   * @param args pairs of argument name and value
   * 
   * @return A JSON object with return messages
   *
   * @since 0.0.1
   */
  @Parameter(name = "x", desc = "List of x co-ordinates of stimuli (degrees).", className = Double.class, min = -90, max = 90, isList = true, defaultValue = "list(0)")
  @Parameter(name = "y", desc = "List of y co-ordinates of stimuli (degrees).", className = Double.class, min = -90, max = 90, isList = true, defaultValue = "list(0)")
  @Parameter(name = "t", desc = "List of stimuli presentation times (ms).", className = Double.class, min = 0, isList = true, defaultValue = "list(200)")
  @Parameter(name = "w", desc = "List of stimuli response windows (ms).", className = Double.class, min = 0, defaultValue = "1500")
  @Parameter(name = "lum", desc = "List of stimuli luminances (cd/m^2).", className = Double.class, min = 0, max = 3183.099, defaultValue = "20")
  @Parameter(name = "size", desc = "Stimulus size (degrees). Can be Goldmann Size I to V (or VI if device has a big wheel)", className = StSize.class, defaultValue = "GV")
  @Parameter(name = "color", desc = "List of stimuli colors.", className = StCol.class, defaultValue = "white")
  @ReturnMsg(name = "error", desc = "Empty string for all good, else error messages from ImoVifa.")
  @ReturnMsg(name = "msg", desc = "JSON Object with all of the other fields described in @ReturnMsg except 'error'.")
  @ReturnMsg(name = "msg.seen", desc = "true if seen, false if not.", className = Boolean.class)
  @ReturnMsg(name = "msg.time", desc = "Response time from stimulus onset if button pressed, -1 otherwise (ms).", className = Double.class, min = -1)
  @ReturnMsg(name = "msg.eyex", desc = "x co-ordinates of pupil at times eyet (degrees).", className = Double.class, isList = true)
  @ReturnMsg(name = "msg.eyey", desc = "y co-ordinates of pupil at times eyet (degrees).", className = Double.class, isList = true)
  @ReturnMsg(name = "msg.eyed", desc = "Diameter of pupil at times eyet (degrees).", className = Double.class, isList = true)
  @ReturnMsg(name = "msg.eyet", desc = "Time of (eyex,eyey) pupil relative to stimulus onset t=0 (ms).", className = Double.class, isList = true)
  @ReturnMsg(name = "msg.jovp", desc = "Any JOVP-specific messages that the JOVP sent back.")
  public MessageProcessor.Packet present(HashMap<String, Object> args) {
    return new MessageProcessor.Packet("");
  }

  /**
   * opiClose: Close OPI connection
   * 
   * @param args pairs of argument name and value
   *
   * @return A JSON object with return messages
   *
   * @since 0.0.1
   */
  @ReturnMsg(name = "error", desc = "Empty string for all good, else error messages from Imo.")
  @ReturnMsg(name = "msg", desc = "JSON Object with all of the other fields described in @ReturnMsg except 'error'.")
  @ReturnMsg(name = "msg.jovp", desc = "Any messages that the JOVP sent back.")
  public MessageProcessor.Packet close() {
    setIsInitialised(false);
    return new MessageProcessor.Packet(true, "");
  }

}

File: opi-rgen/src/main/OPI/R/O900.R
# Open Perimetry Interface implementation for O900
#
# Copyright [2022] [Andrew Turpin & Ivan Marin-Franch]
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require(rjson)

    # environment for this machine in R
if (exists(".opi_env") && !exists("O900", where = .opi_env))
    assign("O900", new.env(), envir = .opi_env)

#' Implementation of opiInitialise for the O900 machine.
#'
#' This is for internal use only. Use [opiInitialise()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#' @param ip IP Address of the perimeter.
#' @param port TCP port of the perimeter.
#' @param ip_Monitor IP Address of the O900 server.
#' @param port_Monitor TCP port of the O900 server.
#' @param eye Eye to set.
#' @param eyeSuite Path to EyeSuite.
#' @param gazeFeed Path where to save gaze feed. Directory must exists
#' @param bigWheel Whether O900 has a big wheel for displaying Goldmann Size VI
#'                    stimuli.
#' @param pres Volume for auditory feedback when a stimulus is presented: 0
#'                means no buzzer.
#' @param resp Volume for auditory feedback when observer presses the clicker: 0
#'                means no buzzer.
#' @param max10000 Whether O900 can handle a maximum luminance of 10000
#'                    apostilbs instead of 4000. Check the settings in EyeSuite
#'
#' @return a list contianing:
#'  * error Empty string for all good, else error messages from Imo.
#'  * msg JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - msg$jovp Any messages that the JOVP sent back.
#'
#' @examples
#' chooseOpi("O900")
#' result <- opiInitialise(ip = "192.126.0.1", port = 50000,
#'                         ip_Monitor = "localhost", port_Monitor = 50001,
#'                         eye = "left", eyeSuite = "C:/XXX/eyeSuite/",
#'                         gazeFeed = "C:/XXX/gazeFeed/", bigWheel = 0, pres = 0,
#'                         resp = 0, max10000 = 0)
#'
#' @seealso [opiInitialise()]
#'
opiInitialise_for_O900 <- function(ip = NULL, port = NULL, ip_Monitor = NULL, port_Monitor = NULL, eye = NULL, eyeSuite = NULL, gazeFeed = NULL, bigWheel = NULL, pres = NULL, resp = NULL, max10000 = NULL) {
    .opi_env$O900$socket <<- open_socket(ip_Monitor, port_Monitor)
    msg <- list(ip = ip, port = port, ip_Monitor = ip_Monitor, port_Monitor = port_Monitor, eye = eye, eyeSuite = eyeSuite, gazeFeed = gazeFeed, bigWheel = bigWheel, pres = pres, resp = resp, max10000 = max10000);
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$O900$socket)

    res <- rjson::fromJSON(readLines(.opi_env$O900$socket, n=1))
    return(res)
}

#' Implementation of opiPresent for the O900 machine.
#'
#' This is for internal use only. Use [opiPresent()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#' @param x List of x co-ordinates of stimuli (degrees).
#' @param y List of y co-ordinates of stimuli (degrees).
#' @param t List of stimuli presentation times (ms).
#' @param w List of stimuli response windows (ms).
#' @param lum List of stimuli luminances (cd/m^2).
#' @param size Stimulus size (degrees). Can be Goldmann Size I to V (or VI if
#'                device has a big wheel)
#' @param color List of stimuli colors.
#'
#' @return a list contianing:
#'  * error Empty string for all good, else error messages from ImoVifa.
#'  * msg JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - msg$seen true if seen, false if not.
#'    - msg$time Response time from stimulus onset if button pressed, -1
#'                  otherwise (ms).
#'    - msg$eyex x co-ordinates of pupil at times eyet (degrees).
#'    - msg$eyey y co-ordinates of pupil at times eyet (degrees).
#'    - msg$eyed Diameter of pupil at times eyet (degrees).
#'    - msg$eyet Time of (eyex,eyey) pupil relative to stimulus onset t=0 (ms).
#'    - msg$jovp Any JOVP-specific messages that the JOVP sent back.
#'
#' @examples
#' chooseOpi("O900")
#' result <- opiPresent(stim = list(x = list(0), y = list(0), t = list(200), w = 1500, lum = 20,
#'                      size = "GV", color = "white"))
#'
#' @seealso [opiPresent()]
#'
opiPresent_for_O900 <- function(stim = list(x = NULL, y = NULL, t = NULL, w = NULL, lum = NULL, size = NULL, color = NULL)) {
if(!exists(".opi_env$O900$socket") || is.null(.opi_env$O900$socket))
    stop("Cannot call opiPresent without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list(x = stim$x, y = stim$y, t = stim$t, w = stim$w, lum = stim$lum, size = stim$size, color = stim$color);
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$O900$socket)

    res <- rjson::fromJSON(readLines(.opi_env$O900$socket, n=1))
    return(res)
}

#' Implementation of opiSetup for the O900 machine.
#'
#' This is for internal use only. Use [opiSetup()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#' @param bgLum Background luminance for eye.
#' @param bgCol Background color for eye.
#' @param fixType Fixation target type for eye.
#' @param fixLum Fixation luminance color for eye (from 0% to 100%).
#' @param f310 Whether to use Logitech's F310 controlles
#'
#' @return a list contianing:
#'  * error Empty string for all good, else error messages from ImoVifa.
#'  * msg JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - msg$jovp Any messages that the JOVP sent back.
#'
#' @examples
#' chooseOpi("O900")
#' result <- opiSetup(settings = list(bgLum = "white", bgCol = "white", fixType = "center",
#'                    fixLum = 50, f310 = 0))
#'
#' @seealso [opiSetup()]
#'
opiSetup_for_O900 <- function(settings = list(bgLum = NULL, bgCol = NULL, fixType = NULL, fixLum = NULL, f310 = NULL)) {
if(!exists(".opi_env$O900$socket") || is.null(.opi_env$O900$socket))
    stop("Cannot call opiSetup without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list(bgLum = settings$bgLum, bgCol = settings$bgCol, fixType = settings$fixType, fixLum = settings$fixLum, f310 = settings$f310);
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$O900$socket)

    res <- rjson::fromJSON(readLines(.opi_env$O900$socket, n=1))
    return(res)
}

#' Implementation of opiClose for the O900 machine.
#'
#' This is for internal use only. Use [opiClose()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#'
#'
#' @return a list contianing:
#'  * error Empty string for all good, else error messages from Imo.
#'  * msg JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - msg$jovp Any messages that the JOVP sent back.
#'
#' @examples
#' chooseOpi("O900")
#' result <- opiClose()
#'
#' @seealso [opiClose()]
#'
opiClose_for_O900 <- function() {
if(!exists(".opi_env$O900$socket") || is.null(.opi_env$O900$socket))
    stop("Cannot call opiClose without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list();
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$O900$socket)

    res <- rjson::fromJSON(readLines(.opi_env$O900$socket, n=1))
    return(res)
}

#' Implementation of opiQueryDevice for the O900 machine.
#'
#' This is for internal use only. Use [opiQueryDevice()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#'
#'
#' @return a list contianing:
#'  * error Empty string for all good, else error message.
#'  * msg JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - msg$jovp Any messages that the JOVP sent back.
#'
#' @examples
#' chooseOpi("O900")
#' result <- opiQueryDevice()
#'
#' @seealso [opiQueryDevice()]
#'
opiQueryDevice_for_O900 <- function() {
if(!exists(".opi_env$O900$socket") || is.null(.opi_env$O900$socket))
    stop("Cannot call opiQueryDevice without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list();
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$O900$socket)

    res <- rjson::fromJSON(readLines(.opi_env$O900$socket, n=1))
    return(res)
}


#' Set background color and luminance in both eyes.
#' Deprecated for ImoVifa and replaced with [opiSetup()].
#' @usage NULL
#' @seealso [opiSetup()]
opiSetBackground_for_ImoVifa <- function(lum, color, ...) {return("Deprecated")}


